[substitutions] speed up config loading: substitutions pass and `!include` redesign (package refactor part 4)

jpeletier opened the pull request:
# What does this implement/fix?

This PR refactors ESPHome’s substitution system so that **all variable substitutions occur exactly once and in a single pass** (with the exception of resolving definitions of remote packages). As a result, configuration processing becomes dramatically faster: in a medium-sized project with ~25 packages, total config validation and processing time dropped from **~2.8 seconds to ~0.15 seconds**, an **18× speed-up** on a Core i9.

While most projects may be small, with the advent of LVGL and UI design, being able to split your project in files and reload quickly to see the user interface changes immediately is a great improvement to the development user experience.

This PR is the **second** in the series of package loading refactor.

> **This PR has been split into reviewable sub-PRs for easier review:**
> - **Part 1:** [#14917](https://github.com/esphome/esphome/pull/14917) — ConfigContext infrastructure and `add_context()` helper
> - **Part 2:** [#14918](https://github.com/esphome/esphome/pull/14918) — Remove `JinjaStr`; rewrite substitution engine around `ContextVars`
> - **Part 3:** [#15031](https://github.com/esphome/esphome/pull/15031) — Refactor `_substitute_item` into a pure function `substitute()`
> - **Part 4 (this PR):** Package pass integration — wires substitution context through the packages pass
>
- **Followed by** #12213.

## The longer read:

Before this PR, substitutions took place in three locations:
1. During local yaml loading, as part of each `!include` with vars.
2. During each remote package processing with vars.
3. During main configuration `do_substitutions_pass` on the full configuration

Additionally, when packages include other packages, substitution passes occuring higher in the hierarchy had to also incrementally cover package contents that already had been included and substituted, further down the hierarchy, in order to cover variables defined further up which resulted in slowdown. For example, consider this package hierarchy: 
```
root:
└─ package1
   ├── package2
   │   └── package4
   │       ├── package5
   │       └── package6
   │           └── package7
   └──  package3
       └── package8
           └── package9
```
Substitutions would happen as follows:

1. Substitute vars in `package7`
2. Substitute `package6` and check again `package7` (this is due to potentially unresolved variables that are defined in package 6 but may be used in package 7)
3. Substitute vars in `package5`
4. Substitute vars in `package4` and again packages 5, 6 and 7 to deal with unresolved vars as above.
5. Substitute vars in `package2` and again packages 4,5,6,7, etc.
6. Substitute vars in `package9`
7. Substitute vars in `package8` and again `package9`
8. Substitute vars in `package3` and again packages 8 and 9
9. Substitute vars in `package1` and again packages 2, 3, 4, 5, 6, 7, 8, 9
11. After all this, actual package merging takes place and all the packages and their components are merged into one big config.
10. A final substitution pass is done on the whole config to cover substitutions in the root file itself and the merged packages 1,2,3,4,5,6,7,8,9. At this point, any undefined variables will output a warning.

Moreover, expressions inside `!include` vars were not reliably expanded when chained across multiple includes, meaning substitution variables could not be dependably used to define new vars passed to packages (see esphome/esphome#12018).

The approach to solve this problem has been as follows:

1. Prerequisite: refactor package processing so that packages are loaded without merging. Substitutions then take place and packages are merged afterwards. In other words, split package processing in two sub-steps: a) loading from disk/remote repository and b) package merging: For more context, see PR esphome#11605, which must be merged prior to this PR.
2. This PR, which:
 * Remove substitution during `!include` and during remote package loading. Instead, these vars are collected into a context and tagged onto the loaded configuration nodes. Parent configuration nodes carry the context they must be interpreted under.
 * Remove `JinjaStr` and its deferred-evaluation machinery. `JinjaStr` held a reference to an unevaluated Jinja string. These were a workaround to be able to continue processing a package containing expressions referencing variables that were not accesible yet.
 * In a similar manner, remove the concept of `ignore_missing`. This flag was necessary because during intermediate `!include`-type substitutions, attempts to substitute a variable may fail due to the variable not being available yet, for example a top-level substitution referenced by an inner package, or an inner package referencing an outer one.
 * During the single substitutions pass, when a tagged node is encountered, its context vars are overlaid and used for evaluating both substitutions and Jinja expressions.
 * Use one persistent Jinja environment instead of creating a new one for every pass.
 * All existing tests pass untouched, plus new aggressive ones to make sure variables are evaluated correctly.

Thus, the new process looks as follows:

1.- Load all remote and local packages; record their context vars.
2.- Perform one and only one substitution pass, with all variables visible.
3.- Merge packages
4.- Resolve `!extend` and `!remove` as before
 
## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes esphome/esphome#12018
- fixes esphome/esphome#9978
- fixes esphome/esphome#12141
- fixes esphome/feature-requests#2473
- fixes esphome/esphome#11998
- fixes esphome/issues#4939
- fixes esphome/esphome#12269
- fixes esphome/esphome#12558
- fixes esphome/esphome#12559
- fixes esphome/esphome#13440

## Test Environment

- [x] all (config)

## Example entry for `config.yaml`:

See example in esphome/esphome#12018, which now works correctly after this fix.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -27,6 +32,12 @@
 DOMAIN = CONF_PACKAGES
 
 
+def validate_has_jinja(value):
```
missing typing

jpeletier replied:
fixed. thanks

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -226,32 +239,114 @@ def get_packages(files) -> dict:
+    return config
+
+
+def merge_packages(config: dict):
```
missing return typing

jpeletier replied:
fixed. thanks

bdraco commented on esphome/components/substitutions/jinja.py:
```diff
@@ -168,7 +140,7 @@ class Jinja(jinja.Environment):
     concat = staticmethod(_concat_nodes_override)
 
-    def __init__(self, context_vars: dict):
+    def __init__(self):
```
```suggestion
    def __init__(self) -> None:
```

jpeletier replied:
fixed. thanks

bdraco commented on esphome/components/substitutions/jinja.py:
```diff
@@ -183,49 +155,32 @@ def __init__(self, context_vars: dict):
+        self.globals = ChainMap(
+            SAFE_GLOBALS,
+            {**self.globals},
+        )
```
Is the ChainMap faster? Can you benchmark this?

jpeletier replied:
It is not faster. Just a remnant of an intermediate iteration. Fixed.

bdraco commented on esphome/yaml_util.py:
```diff
@@ -87,6 +88,38 @@ def make_data_base(
+            **(context_vars or {}),
+        }
+
+    if isinstance(value, ConfigContext):
+        value.set_context({**value.vars, **(context_vars or {})})
+        return value
```
Missing coverage here

jpeletier replied:
Added unit tests to cover this fully

bdraco commented on esphome/yaml_util.py:
```diff
@@ -87,6 +88,38 @@ def make_data_base(
 
+class ConfigContext:
+    """Holds substitution vars that should be applied
+    to the tagged node and its children."""
```
Please expand the doc string to explain how this works

jpeletier replied:
done, thanks.

bdraco commented on esphome/components/substitutions/jinja.py:
```diff
@@ -168,7 +140,7 @@ class Jinja(jinja.Environment):
     concat = staticmethod(_concat_nodes_override)
 
-    def __init__(self, context_vars: dict):
+    def __init__(self):
```
```suggestion
    def __init__(self) -> None:
```

jpeletier replied:
fixed. thanks

bdraco commented on esphome/components/substitutions/__init__.py:
```diff
@@ -157,63 +170,91 @@ def _expand_substitutions(
+    parent_context: ContextVars,
+    strict_undefined: bool,
 ) -> Any | None:
+    """Applies substitutions in-place to the given item.
```
This function is a bit hard to follow. It could use some comments

bdraco commented on esphome/components/substitutions/__init__.py:
```diff
@@ -6,13 +7,22 @@
 _LOGGER = logging.getLogger(__name__)
 
+ContextVars = ChainMap[str, Any]
+jinja = Jinja()
```
TODO (review note to self): adds a package global.. verify that we don't store state here

jpeletier replied:
It is intended to just be a "typedef", hopefully I am doing it the right way

bdraco replied:
This looks like a module level mutable global.

If it stores state it should live in CORE.data instead so it gets cleared with reset

bdraco replied:
It looks like it won't be a problem between runs, but we should document it and explain why its ok

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -226,32 +239,114 @@ def get_packages(files) -> dict:
+    return config
+
+
+def merge_packages(config: dict):
```
nit: missing return type

jpeletier replied:
Fixed -thanks

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -226,32 +239,114 @@ def get_packages(files) -> dict:
+    def process_package_callback(package_config, context):
+        nonlocal config
+        config = merge_config(package_config, config)
+        return _walk_packages(package_config, process_package_callback)
```
missing typing.

Would be nicer to write this without nonlocal. Maybe in a class. The callback mutation pattern makes the code harder to follow

jpeletier replied:
refactored `nonlocal` out.

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -226,32 +239,114 @@ def get_packages(files) -> dict:
+        return config
+    substitutions = config.pop(CONF_SUBSTITUTIONS, {})
+
+    def process_package_callback(package_config, context_vars):
```
same comment as https://github.com/esphome/esphome/pull/12126/files#r2585836441

jpeletier replied:
refactored nonlocal out.

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -27,6 +32,12 @@
 DOMAIN = CONF_PACKAGES
 
 
+def validate_has_jinja(value):
```
nit: missing typing

jpeletier replied:
Fixed - thanks

jpeletier commented:
Thanks for the review!

> This PR is quite large and invasive. 

Yes it looks very large, it is untying a big knot with how variables are handled everywhere. This PR is part 2. Perhaps you could look at part 1 #11605 first, which should make things easier to follow.

> I'm a bit concerned that the test coverage isn't sufficient to cover the scope of it.

Prior PRs already merged have aimed to increase substitutions tests coverage #12210, #12158, #12145, #11611, #11584, #10814, #9469, #8955 and #11605 itself so that packages and substitutions tests have to pass before and after the PR we are reviewing now. This PR adds more tests, but does not modify existing tests, like playing by a test-driven development set of rules; the new algorithm must produce the same result.

Additionally, I have a big monster project that I use to make sure it works; tests added come from situations in there, but I run it on my project anyway.

Anyway, thank your for your feedback!. I will be addressing your review comments now. If you can think of a specific list of cases you think we're not covering, I am more than happy to add them.

bdraco commented:
@jpeletier please resolve conflicts when you have a free moment.

Thanks!

bdraco commented on esphome/config.py:
```diff
@@ -998,17 +999,15 @@ def validate_config(
-            result.add_error(err)
-            return result
+    try:
+        substitutions.do_substitution_pass(config, command_line_substitutions)
+    except vol.Invalid as err:
+        result.add_error(err)
+        return result
```
Missing coverage for these lines

jpeletier replied:
fixed. thanks!

bdraco commented on esphome/components/substitutions/__init__.py:
```diff
@@ -6,13 +7,23 @@
 
+ContextVars = ChainMap[str, Any]
+SubstitutionPath = list[int | str]
+ErrList = list[(UndefinedError, SubstitutionPath, str)]
```
```suggestion
ErrList = list[tuple[UndefinedError, SubstitutionPath, str]] 
```

bdraco commented on esphome/components/substitutions/__init__.py:
```diff
@@ -230,10 +323,27 @@ def do_substitution_pass(
+                expression,
+                err.message,
+                location,
+            )
```
It woul  be a bit cleaner to make this a helper to show intent

jpeletier replied:
done, thanks

bdraco commented:
The complexity here is making the review take a lot of time. I asked claude to build a diagram of how this works.

In case its helpful:

```
                                                                                                                                                                                              
  ┌─────────────────────────────────────────────────────────────────────────────┐                                                                                                                                             
  │                         CONFIGURATION LOADING PHASE                          │                                                                                                                                            
  └─────────────────────────────────────────────────────────────────────────────┘                                                                                                                                             
                                                                                                                                                                                                                              
    YAML File                    !include with vars              Remote Package                                                                                                                                               
        │                              │                              │                                                                                                                                                       
        ▼                              ▼                              ▼                                                                                                                                                       
    load_yaml()              construct_include()            _process_remote_package()                                                                                                                                         
        │                              │                              │                                                                                                                                                       
        │                              ▼                              ▼                                                                                                                                                       
        │                      add_context(result, vars)    add_context(pkg, vars)                                                                                                                                            
        │                              │                              │                                                                                                                                                       
        │                              ▼                              ▼                                                                                                                                                       
        │                    ┌─────────────────────────────────────────┐                                                                                                                                                      
        │                    │  ConfigContext mixin applied to node   │                                                                                                                                                       
        │                    │  node._context_vars = {var1: val1, ...}│                                                                                                                                                       
        │                    └─────────────────────────────────────────┘                                                                                                                                                      
        │                              │                              │                                                                                                                                                       
        └──────────────────────────────┴──────────────────────────────┘                                                                                                                                                       
                                       │                                                                                                                                                                                      
                                       ▼                                                                                                                                                                                      
                          Config tree with tagged nodes                                                                                                                                                                       
                          (vars NOT yet substituted)                                                                                                                                                                          
                                                                                                                                                                                                                              
                                                                                                                                                                                                                              
  ┌─────────────────────────────────────────────────────────────────────────────┐                                                                                                                                             
  │                         SUBSTITUTION PASS (single pass)                      │                                                                                                                                            
  └─────────────────────────────────────────────────────────────────────────────┘                                                                                                                                             
                                                                                                                                                                                                                              
    do_substitution_pass(config, cmd_line_subs)                                                                                                                                                                               
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    ┌─────────────────────────────────────────┐                                                                                                                                                                               
    │ Extract top-level substitutions:        │                                                                                                                                                                               
    │ substitutions = config[CONF_SUBSTITUTIONS] + cmd_line_subs                                                                                                                                                              
    └─────────────────────────────────────────┘                                                                                                                                                                               
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    _push_context(substitutions, ChainMap())                                                                                                                                                                                  
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    ┌─────────────────────────────────────────┐                                                                                                                                                                               
    │ Creates root context:                   │                                                                                                                                                                               
    │ ChainMap({a: 1, b: 2, ...})            │                                                                                                                                                                                
    └─────────────────────────────────────────┘                                                                                                                                                                               
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    substitute(config, [], root_context, ...)                                                                                                                                                                                 
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    ┌─────────────────────────────────────────────────────────────────────────┐                                                                                                                                               
    │                        RECURSIVE TREE WALK                               │                                                                                                                                              
    │                                                                          │                                                                                                                                              
    │   For each node in config tree:                                          │                                                                                                                                              
    │                                                                          │                                                                                                                                              
    │   1. push_context(node, parent_context)                                  │                                                                                                                                              
    │      │                                                                   │                                                                                                                                              
    │      ▼                                                                   │                                                                                                                                              
    │      Is node a ConfigContext? ──No──► return parent_context              │                                                                                                                                              
    │      │                                                                   │                                                                                                                                              
    │      Yes                                                                 │                                                                                                                                              
    │      │                                                                   │                                                                                                                                              
    │      ▼                                                                   │                                                                                                                                              
    │      _push_context(node.vars, parent_context)                            │                                                                                                                                              
    │      │                                                                   │                                                                                                                                              
    │      ▼                                                                   │                                                                                                                                              
    │      ┌───────────────────────────────────────┐                           │                                                                                                                                              
    │      │ ChainMap creates NEW CHILD scope:     │                           │                                                                                                                                              
    │      │                                       │                           │                                                                                                                                              
    │      │   parent: {a: 1, b: 2}               │                           │                                                                                                                                               
    │      │      ▲                                │                           │                                                                                                                                              
    │      │      │ (falls through on miss)       │                           │                                                                                                                                               
    │      │   child:  {c: 3, a: 99}  ◄─ overrides│                           │                                                                                                                                               
    │      │                                       │                           │                                                                                                                                              
    │      │ Lookup "a" → 99 (from child)         │                           │                                                                                                                                               
    │      │ Lookup "b" → 2  (from parent)        │                           │                                                                                                                                               
    │      └───────────────────────────────────────┘                           │                                                                                                                                              
    │                                                                          │                                                                                                                                              
    │   2. Process node with this context:                                     │                                                                                                                                              
    │      • dict  → recurse on keys & values                                  │                                                                                                                                              
    │      • list  → recurse on elements                                       │                                                                                                                                              
    │      • str   → _expand_substitutions()                                   │                                                                                                                                              
    │                                                                          │                                                                                                                                              
    └─────────────────────────────────────────────────────────────────────────┘                                                                                                                                               
                                                                                                                                                                                                                              
                                                                                                                                                                                                                              
  ┌─────────────────────────────────────────────────────────────────────────────┐                                                                                                                                             
  │                      VARIABLE EXPANSION (_expand_substitutions)              │                                                                                                                                            
  └─────────────────────────────────────────────────────────────────────────────┘                                                                                                                                             
                                                                                                                                                                                                                              
    Input: "Hello ${name}, pin ${pin+1}"                                                                                                                                                                                      
    Context: ChainMap({name: "World", pin: 12})                                                                                                                                                                               
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    Find ${name} ──► context_vars.get("name") ──► "World"                                                                                                                                                                     
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    "Hello World, pin ${pin+1}"                                                                                                                                                                                               
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    No more $vars, but has Jinja ${...+...}                                                                                                                                                                                   
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    _expand_jinja() ──► jinja.expand("...", context_vars)                                                                                                                                                                     
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    "Hello World, pin 13"                                                                                                                                                                                                     
                                                                                                                                                                                                                              
                                                                                                                                                                                                                              
  ┌─────────────────────────────────────────────────────────────────────────────┐                                                                                                                                             
  │                   RECURSIVE VARIABLE RESOLUTION (_push_context)              │                                                                                                                                            
  └─────────────────────────────────────────────────────────────────────────────┘                                                                                                                                             
                                                                                                                                                                                                                              
    Problem: Variables can reference other variables defined later                                                                                                                                                            
                                                                                                                                                                                                                              
    substitutions:                                                                                                                                                                                                            
      b: ${a + 1}     # depends on 'a'                                                                                                                                                                                        
      a: 10           # defined after 'b'                                                                                                                                                                                     
      c: ${b * 2}     # depends on 'b'                                                                                                                                                                                        
                                                                                                                                                                                                                              
    Solution: Resolver callback + on-demand resolution                                                                                                                                                                        
                                                                                                                                                                                                                              
    _push_context(local_vars, parent):                                                                                                                                                                                        
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    unresolved = {b: "${a+1}", a: 10, c: "${b*2}"}                                                                                                                                                                            
    local_vars = {}  (empty, will fill as resolved)                                                                                                                                                                           
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    resolver_context[Resolver] = resolve  ◄── callback registered                                                                                                                                                             
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    while unresolved:                                                                                                                                                                                                         
        resolve(next key)  ──► resolve("b")                                                                                                                                                                                   
                                    │                                                                                                                                                                                         
                                    ▼                                                                                                                                                                                         
                               substitute("${a+1}", resolver_context)                                                                                                                                                         
                                    │                                                                                                                                                                                         
                                    ▼                                                                                                                                                                                         
                               Lookup "a" → Missing                                                                                                                                                                           
                                    │                                                                                                                                                                                         
                                    ▼                                                                                                                                                                                         
                               Call Resolver("a") ◄── recursive!                                                                                                                                                              
                                    │                                                                                                                                                                                         
                                    ▼                                                                                                                                                                                         
                               resolve("a") → 10                                                                                                                                                                              
                               local_vars["a"] = 10                                                                                                                                                                           
                                    │                                                                                                                                                                                         
                                    ▼                                                                                                                                                                                         
                               Return to "${a+1}" → 11                                                                                                                                                                        
                               local_vars["b"] = 11                                                                                                                                                                           
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    resolve("c") → "${b*2}" → 22                                                                                                                                                                                              
    local_vars["c"] = 22                                                                                                                                                                                                      
        │                                                                                                                                                                                                                     
        ▼                                                                                                                                                                                                                     
    Final: ChainMap({a: 10, b: 11, c: 22}, parent)                                      
```

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -285,6 +294,25 @@ def _walk_packages(
+    elif isinstance(package_config, str):
+        # If a string is found, it may be a potential github:// shorthand
+        # try to substitute it and then validate it actually is a remote package definition
+        result = substitute(package_config, [], context_vars, False) or package_config
```
I'd pass the args to substitute as kwargs so its a bit more readable

jpeletier replied:
fixed. thanks!

bdraco commented on esphome/components/packages/__init__.py:
```diff
@@ -294,20 +322,34 @@ def do_packages_pass(config: dict, skip_update: bool = False) -> dict:
+        if CONF_PACKAGES not in package_config:
+            # This package has no nested packages, so we're done.
+            return package_config
+        context_vars = push_context(package_config, context_vars)
+        context_vars = push_context(package_config[CONF_PACKAGES], context_vars)
```
Why are both `package_config` and `package_config[CONF_PACKAGES]` checked for context? Under what circumstances would the packages key itself have a ConfigContext mixin? Please add a comment explaining this or simplify if only one is needed.

jpeletier replied:
```python
context_vars = push_context(package_config, context_vars)
```
This is because `package_config` may come from a previous `!include` and therefore contain vars in the root.



```python
context_vars = push_context(package_config[CONF_PACKAGES], context_vars)
```
This is because the user could've done:

```yaml
packages: !include 
  file: package_list.yaml
  vars:
    a: 7
    b: 3
```

where `package_list.yaml` could be just a plain list or dict:

```yaml
- !include package1.yaml # note: `a` (with a=7) and `b` are visible in package1.yaml.
- !include
    file: package2.yaml # `b` and `c` (c=4) and `a` (with a=8) are visible in package2.yaml
    vars:
      c: ${b+1}
      a: ${a+1}
```
Since you can put `!include` everywhere, and the referenced file can be any valid YAML, this could happen.

Note that in this process we are only traversing the `packages` lineage in the tree, that is why we "hop twice" in a row.

Added some comments to clarify.

bdraco commented:
This approach of dynamically mixing context into YAML nodes is similar to what Home Assistant's [annotatedyaml](https://github.com/home-assistant-libs/annotatedyaml) library does. Have we considered whether there's potential for shared infrastructure, or is ESPHome's use case different enough to warrant a separate implementation? Adding another mixin layer increases complexity.

jpeletier commented:
I took a look at annotatedyaml, I didn't know about it. 

If I understood correctly, in essence, it replaces str,dict,list with custom classes that admit metadata, which would be useful to carry context vars and "ESPHomeDataBase" stuff.

I think that yes there is potential to share, though most of the work would happen in ESPHome's `yaml_util` module, which would need to be converted to depend on annotatedyaml, though we'd need to review if annotatedyaml's custom constructors for `!include`, `!secret`, `!include_dir_list*`, etc, conflict or can be overriden by ESPHome's own.

In any case, the purpose of this PR is different and a potential migration to annotatedyaml could be done afterwards without major pain, context vars would be another slot in the custom classes for example.

djtef commented:
Hi @jpeletier,

First of all, congrats on the recent commits — this is a *huge* step forward.  
The redesign makes a lot of sense, and the added tests are very reassuring. Thanks again for the work you’re putting into this.

If I understand the new model correctly, package **definitions** (including remote packages and their `files:` entries) can now rely on substitution variables that are progressively consolidated while resolving packages.

This makes me wonder whether it is now supported (or expected to work) to use substitutions and simple expressions directly in remote package file paths.

For example, would something like this now be valid?

```yaml
substitutions:
  type: aquaair
  router: yes
  network: wifi

packages:
  remote_package:
    url: https://github.com/djtef/toug
    ref: main
    refresh: 1s
    files:
      - esphome/${type}${router == "yes" and "_router" or ""}_${network}.yaml
```
This would allow replacing the previous “comment/uncomment one file” approach with a fully dynamic selection, which would be really nice for modular projects.

Before I rely on this pattern, I just wanted to confirm:

- Is substitution (and simple expressions) in files: paths now supported by the new package resolution logic?
- Or is this something that is still planned for a later step (e.g. part 3 of the package refactor)?

If helpful, I’d be more than happy to test this against my real-world project and provide feedback.

Thanks again — this work will make complex, reusable ESPHome setups much easier to maintain.

jpeletier commented on esphome/components/substitutions/__init__.py:
```diff
@@ -157,67 +171,157 @@ def _expand_substitutions(
+            new_k = substitute(k, path + [k], context_vars, strict_undefined, errors)
+            v = substitute(v, path + [k], context_vars, strict_undefined, errors)
+            if str(new_k) != str(k):
+                result[new_k] = merge_config(item.get(k), v)
```
solved and added tests

tanishqmanuja commented:
Since this is not a `priority for them`. So is there any workaround for now @jpeletier ?
I am too annoyed from https://github.com/esphome/esphome/issues/12141 🥲

tanishqmanuja commented:
@edwardtfn didn't work with include when i tested last. Had to write the package as a substitution variable unfortunately for now.

But if you include this package in another one that will work ( meaning if you define substitutions in another file and include this one )

jpeletier commented:
> Would this PR allow something like:
> 
> ```yaml
> substitutions:
>   config_type: private
>   overrides: !include ${config_type}.yaml
> ```
> 
> ?
> 
> I have seen a similar syntax in forum posts, but cannot get it to work. The `!include` simply tries to include a file named `${config_type}.yaml`, rather than resolving the variable first.

No, but the next one will:
https://github.com/esphome/esphome/pull/12213

jpeletier committed:
refactor packages

bdraco committed:
reduce nesting, improve readability, add comments

bdraco committed:
typing

bdraco committed:
split up _walk_packages

bdraco committed:
reduce nesting

bdraco committed:
tweak style

bdraco committed:
split

bdraco committed:
touch ups

bdraco committed:
touch ups

bdraco committed:
tidy up

bdraco committed:
touch ups

bdraco committed:
refactor flow to make it easier to follow

bdraco committed:
typing

bdraco committed:
missing coverage

bdraco committed:
fix contextvars types

bdraco committed:
make a class for do_packages_pass

bdraco committed:
some more typing / none fixes

bdraco committed:
make do_packages_pass kwargs only to force improving readability

bdraco committed:
test deprecation

bdraco committed:
lint, docs

bdraco committed:
comments

bdraco committed:
handle none case correctly

bdraco committed:
lint

jpeletier committed:
use CONF_PACKAGES

jpeletier committed:
reorder functions so dependencies appear earlier
